docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,317 @@
+* FR3: CAPV MUST support static credentials (Story 3)
+* FR4: CAPV MUST support clusterctl move scenarios
+* FR5: CAPV MUST prevent privilege escalation allowing users to create clusters in accounts they should not be able to (Story 2)
+
+#### Non-Functional Requirements
+
+* NFR1: Unit tests MUST exist for cluster and machine controllers that utilize the credentials
+* NFR2: e2e tests MUST exist for multi account scenarios
+
+### Implementation Details/Notes/Constraints
+
+#### Current State
+
+CAPV currently uses a session manager to create and cache sessions. Sessions are created (or retrieved from the cache) during cluster and vSphere VM reconcile loops.
+vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.
+
+```go
+type VMContext struct {
+    *ControllerContext
+    VSphereVM   *v1alpha3.VSphereVM
+    PatchHelper *patch.Helper
+    Logger      logr.Logger
+    Session     *session.Session
+}
+```
+
+#### Proposed Changes
+
+The proposed changes below allow users to specify the account to use for clusters during runtime. The credentials can be provided by two approaches:
+
+* Referencing a cluster scoped `VSphereClusterIdentity` to be used for credentials.
+* Referencing a `Secret` that contains the credentials in the same namespace as the cluster.
+
+Changed Resources
+
+* `VSphereCluster`
+
+New Resources
+
+* A cluster-scoped `VSphereClusterIdentity` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.
+
+Changes to VSphereCluster
+A new field is added to the VSphereClusterSpec to reference the `VSphereClusterIdentity`. We intend to use a `VSphereIdentityReference` type, similar to that of `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VSphereClusterIdentity is the account to be used for vcenter actions
+type VSphereClusterIdentity struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VSphereClusterIdentitySpec `json:"spec,omitempty"`
+    Status VSphereClusterIdentityStatus `json:"status, omitempty"`
+}
+
+type VSphereClusterIdentitySpec struct {
+    // secretsRef references a secret in the CAPV controller namespace with the credentials to use
+    secretRef `json:"secretRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using a label selector.
+    // If this object is nil, no namespaces will be allowed.
+    // If this object is empty {}, clusters can use this identity from any namespace.
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
```

Comment:
according the general guidelines being proposed in CAPI, NamespaceList is missing

Reply:
+1

Reply:
The PR calls out to deprecate `NamespaceList` for v1alpha4

`Providers SHOULD mark NamespaceList as deprecated from v1alpha4`